### PR TITLE
NATIVE-239: Store pois offline

### PR DESCRIPTION
**Pull request by TechPiranja**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **NATIVE**.

### CODE REVIEW

**Comment by klinzo** [comment]
:tada:

File: src/modules/endpoint/sagas/__tests__/loadPois.spec.js [hunk 1]
```diff
@@ -44,29 +45,31 @@ describe('loadPois', () => {
     const dataContainer = new DefaultDataContainer()
 
     await runSaga({}, loadPois, city, language, dataContainer, false).toPromise()
-
-    expect(await dataContainer.getPois()).toStrictEqual(mockPois)
+    var test = await dataContainer.getPois(city, language)
+    console.log(test)
```
Comment (klinzo) [inline src/modules/endpoint/sagas/__tests__/loadPois.spec.js:49]: Could be removed, right?

File: src/modules/app/StateType.js [hunk 2]
```diff
@@ -45,6 +45,32 @@ export type CategoryRouteStateType = {|
...
+  +path: ?string, // path is null for the poi-lists route
+  +language: string,
+  +city: string
+|}
+
+export type PoiRouteStateType = {|
+  +status: 'ready',
+  ...PoiRouteConfigType,
+  +models: $ReadOnlyArray<PoiModel>,
+  +allAvailableLanguages: $ReadOnlyMap<string, ?string> // including the current content language
```
Comment (klinzo) [inline src/modules/app/StateType.js:58]: Why do we need `?string` here?

File: src/modules/endpoint/sagas/watchFetchPoi.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        type: 'PUSH_POI',
+        params: { pois, resourceCache, path, cityLanguages, key, language, city }
+      }
+      yield put(insert)
+    } else {
+      const allAvailableLanguages = path === null ? new Map(cityLanguages.map(lng => [lng.code, null])) : null
+      const failed: FetchPoiFailedActionType = {
+        type: 'FETCH_POI_FAILED',
+        params: {
+          message: 'Could not load event.',
```
Comment (klinzo) [inline src/modules/endpoint/sagas/watchFetchPoi.js:39]:
```suggestion
          message: 'Could not load poi.',
```

File: src/modules/app/StateType.js (hunk 2 above)
Comment (TechPiranja) [reply src/modules/app/StateType.js:58]: I saw it in other RouteStateTypes so i added it here too. Not sure if it is needed.

**Comment by TechPiranja** [comment]
> Looking good! But only really testable with UI in a separate PR, right?
Yes, i will open a new PR after merging this one.

File: src/modules/app/StateType.js (hunk 2 above)
Comment (maxammann) [reply src/modules/app/StateType.js:58]: For events we have it because of this snippet:

```
      14│   const getEventRoute = (): EventRouteStateType => {
      15│     if (!path) {
 [1]  16│       const allAvailableLanguages = new Map(cityLanguages.map(lng => [lng.code, null]))
      17│       return { status: 'ready', path: null, models: events, allAvailableLanguages, language, city }
      18│     }
      19│     const event: ?EventModel = events.find(event => event.path === path)
```

File: src/modules/app/StoreActionType.js [hunk 4]
```diff
@@ -1,180 +1,224 @@
...
-  type: 'CLEAR_RESOURCES_AND_CACHE'
-|}
-
-export type StoreActionType =
-  ToggleDarkModeActionType
-  | CitiesActionType
-  | CityContentActionType
-  | SetContentLanguageActionType
-  | ClearResourcesAndCacheActionType
+// @flow
```
Comment (maxammann) [inline src/modules/app/StoreActionType.js:1]: Why did you change the mode of this file from 644 to 755? Please revert this. Else the diff it not easy to check.

File: src/modules/app/StoreActionType.js [hunk 5]
```diff
@@ -1,180 +1,224 @@
...
+    +city: string
+  |}
+|}
+export type FetchPoiFailedActionType = {|
+  type: 'FETCH_POI_FAILED',
+  +params: {|
+    +message: string,
+    +code: ErrorCodeType,
+    +key: string,
+    +allAvailableLanguages: ?$ReadOnlyMap<string, ?string>,
```
Comment (maxammann) [inline src/modules/app/StoreActionType.js:110]:
```suggestion
    +allAvailableLanguages: ?$ReadOnlyMap<string, string>,

```

This should probably be `string`

File: src/modules/endpoint/DataContainer.js [hunk 6]
```diff
@@ -1,115 +1,131 @@
...
-   * Deletes all pages and files that are stored offline in the filesystem
-   */
-  clearOfflineCache (): Promise<void>,
-
-  /**
-   * Clears all in-memory caches
-   */
-  clearInMemoryCache (): void
-}
+// @flow
```
Comment (maxammann) [inline src/modules/endpoint/DataContainer.js:1]: Same for this file. Change the mode back to 644.

If possible rewrite/rebase the history that this change does no longer exist. Else the history shows that you removed the file and readded every line.

File: src/modules/endpoint/DatabaseConnector.js [hunk 7]
```diff
@@ -339,6 +352,46 @@ class DatabaseConnector {
...
+  async storePois (pois: Array<PoiModel>, context: DatabaseContext) {
+    const jsonModels = pois.map((poi: PoiModel): ContentPoiJsonType => ({
+      path: poi.path,
+      title: poi.title,
+      content: poi.content,
+      thumbnail: poi.thumbnail,
+      availableLanguages: poi.availableLanguages,
+      excerpt: poi.excerpt,
+      location: poi.location,
+      lastUpdate: poi.lastUpdate,
```
Comment (maxammann) [inline src/modules/endpoint/DatabaseConnector.js:365]: What is the format of this timestamp? You probably need to store it in ISO8601: `object.lastUpdate.toISOString()`

File: src/modules/endpoint/DatabaseConnector.js [hunk 8]
```diff
@@ -339,6 +352,46 @@ class DatabaseConnector {
...
+    return json.map((jsonObject: ContentPoiJsonType) => {
+      return new PoiModel({
+        path: jsonObject.path,
+        title: jsonObject.title,
+        content: jsonObject.content,
+        thumbnail: jsonObject.thumbnail,
+        availableLanguages: jsonObject.availableLanguages,
+        excerpt: jsonObject.excerpt,
+        location: jsonObject.location,
+        lastUpdate: jsonObject.lastUpdate,
```
Comment (maxammann) [inline src/modules/endpoint/DatabaseConnector.js:390]: `lastUpdate: moment(jsonObject.last_update, moment.ISO_8601),`

Please check that load/store code actually works. You can do this using `adb shell`. Then you can use cat to see the jsons: `cat /data/data/de.tuerantuer.integreat/files/cities.json`

File: src/modules/endpoint/reducers/pushPoi.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+const pushPoi = (state: CityContentStateType, action: PushPoiActionType): CityContentStateType => {
+  const { pois, path, key, language, resourceCache, cityLanguages, city } = action.params
+
+  if (!key) {
+    throw new Error('You need to specify a key!')
+  }
+
+  const getPoiRoute = (): PoiRouteStateType => {
+    const allAvailableLanguages = new Map(cityLanguages.map(lng => [lng.code, null]))
```
Comment (maxammann) [inline src/modules/endpoint/reducers/pushPoi.js:15]: Here is the reason why the `availableLanguages` state has a `?string`. Is this required here? Can we link to pois in different languages? @Schedulaar do you know more here?

File: src/testing/builder/PoiModelBuilder.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
+// @flow
+
+import { LocationModel, PoiModel } from '@integreat-app/integreat-api-client'
+import moment from 'moment'
+
+const map = {}
```
Comment (maxammann) [inline src/testing/builder/PoiModelBuilder.js:6]:
```suggestion
const map = {ID1: 'Farm1', ID2: 'Farm2'}
```

File: src/modules/app/StateType.js (hunk 2 above)
Comment (Schedulaar) [reply src/modules/app/StateType.js:58]: For the events, if a path is falsy, we interpret it as the `overview` / `list` view of all events.

File: src/modules/endpoint/reducers/pushPoi.js (hunk 9 above)
Comment (Schedulaar) [reply src/modules/endpoint/reducers/pushPoi.js:15]: We should be able to do so. As for the events, this should only filled with `null` if it corresponds to the list view.

File: src/modules/app/StateType.js (hunk 2 above)
Comment (maxammann) [reply src/modules/app/StateType.js:58]: @TechPiranja can you add a note to the StateType definitions.

```
type LanguageKey = string
type Path = ?string // Path can be falsy if the current displayed view is a list of events or POIs
type AllAvailableLanguagesType = $ReadOnlyMap<LanguageKey, Path>
```

Then reuse these definitions in the StateType:
```
+allAvailableLanguages: AllAvailableLanguagesType
```

File: src/testing/builder/PoiModelBuilder.js [hunk 11]
```diff
@@ -3,9 +3,7 @@
 import { LocationModel, PoiModel } from '@integreat-app/integreat-api-client'
 import moment from 'moment'
 
-const map = {}
-map.ID1 = 'Farm1'
-map.ID2 = 'Farm2'
+const map = {ID1: 'Farm1', ID2: 'Farm2'}
```
Comment (Schedulaar) [inline src/testing/builder/PoiModelBuilder.js:6]:
```suggestion
const availableLanguages = {de: '/de/test', en: '/en/test'}
```

File: src/modules/app/StateType.js [hunk 12]
```diff
@@ -47,6 +48,35 @@ export type CategoryRouteStateType = {|
...
 
+export type PoiRouteConfigType = {|
+  +path: ?string, // path is null for the poi-lists route
+  +language: string,
+  +city: string
+|}
+
+type LanguageKey = string
+type Path = ?string // Path can be falsy if the current displayed view is a list of events or POIs
+type AllAvailableLanguagesType = $ReadOnlyMap<LanguageKey, Path>
```
Comment (klinzo) [inline src/modules/app/StateType.js:59]: I'd rather inline `Path` and `LanguageKey`. Not sure if a type for `AllAvailableLanguageType` is already declared somewehre else. If not, maybe better to just inline it completely for consistency.

File: src/modules/endpoint/DataContainer.js (hunk 6 above)
Comment (klinzo) [reply src/modules/endpoint/DataContainer.js:1]: Still relevant.

File: src/modules/endpoint/reducers/morphContentLanguage.js [hunk 13]
```diff
@@ -102,11 +102,59 @@ const eventRouteTranslator = (newEvents: $ReadOnlyArray<EventModel>, newLanguage
...
+      return {
+        status: 'languageNotAvailable',
+        allAvailableLanguages,
+        language: newLanguage,
+        city
+      }
+    }
+
+    const translatedPath = allAvailableLanguages.get(newLanguage)
+    if (!translatedPath) { // Route is a list of all events
```
Comment (klinzo) [inline src/modules/endpoint/reducers/morphContentLanguage.js:124]:
```suggestion
    if (!translatedPath) { // Route is a list of all pois
```

File: src/modules/app/StateType.js (hunk 12 above)
Comment (maxammann) [reply src/modules/app/StateType.js:59]: I suggested to do it this way because it was very unclear why Path can be optional. therefore that comment was added

File: src/modules/endpoint/DatabaseConnector.js [hunk 14]
```diff
@@ -82,6 +83,18 @@ type ContentCityJsonType = {|
   tunewsEnabled: boolean
 |}
 
+type ContentPoiJsonType = {|
```
Comment (maxammann) [inline src/modules/endpoint/DatabaseConnector.js:86]: A JsonType should not have non-json types:
* LocaltionModel
* Moment
* Map<string, string>

File: src/modules/endpoint/DatabaseConnector.js [hunk 15]
```diff
@@ -381,15 +393,28 @@ class DatabaseConnector {
...
         title: jsonObject.title,
         content: jsonObject.content,
         thumbnail: jsonObject.thumbnail,
-        availableLanguages: jsonObject.availableLanguages,
+        availableLanguages,
         excerpt: jsonObject.excerpt,
-        location: jsonObject.location,
-        lastUpdate: moment(jsonObject.last_update, moment.ISO_8601),
+        location: new LocationModel({
+          name: null, // todo: NATIVE-549
```
Comment (maxammann) [inline src/modules/endpoint/DatabaseConnector.js:407]: @klinzo should we disable the fetching of events until 549 is done? Maybe just comment out the fetching/persisting code?

Comment (klinzo) [reply src/modules/endpoint/DatabaseConnector.js:407]: Do you mean disabling fetching of pois? @TechPiranja is fetching of pois initiated anywhere right now with this PR?

Comment (TechPiranja) [reply src/modules/endpoint/DatabaseConnector.js:407]: @klinzo  no, the pois are not being fetched currently.

It **would have been** fetched if i added the "FETCH_POI" action type inside the cityContentReducer.
( which is currently not the case, but i stashed that code for the Poi UI Ticket )

Comment (klinzo) [reply src/modules/endpoint/DatabaseConnector.js]: @maxammann so I guess we can merge this, right?

File: src/modules/endpoint/sagas/loadPois.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+  if (poisAvailable && !forceRefresh) {
+    try {
+      console.debug('Using cached pois')
+      return yield call(dataContainer.getPois, city, language)
+    } catch (e) {
+      console.warn('An error occurred while loading pois from JSON', e)
+    }
+  }
+
```
Comment (klinzo) [inline src/modules/endpoint/sagas/loadPois.js:25]: A debug log would be nice here.
```suggestion

  console.debug('Fetching pois')

```
